tests/app/dao/test_failed_login_dao.py

```diff
@@ -0,0 +1,82 @@
+from datetime import datetime, timedelta
+
+import pytest
+
+from app.dao.failed_logins_by_ip_dao import (
+    dao_create_failed_login_for_ip,
+    dao_get_failed_logins,
+    dao_get_latest_failed_login_by_ip,
+)
+from app.errors import InvalidRequest
+from app.failed_logins_by_ip.rest import check_throttle_for_ip
+from app.models import FailedLoginCountByIP
+from tests.app.db import create_failed_login
+
+
+def test_get_failed_logins_returns_all_failed_logins(notify_db_session):
+    """
+    Creates 3 failed login records and checks that these are returned by dao_get_failed_logins.
+    """
```

Comment:
define IP in one place, like

test_ip = "192.0.2.15"

then use throughout the rest of the method - it'll make it easier if it ever needs to be changed